repl/convert_ast.py

```diff
@@ -7,19 +7,68 @@
 import qwerty_ast
 #################### COMMON CODE FOR BOTH @QPU AND @CLASSICAL DSLs ####################
 
+# quick fix, going to create another file most likely for errors
+class QwertySyntaxError(Exception):
+    pass
+
 def convert_to_qwerty(py_ast):
     print(ast.dump(py_ast, indent=4))
-    value = unpack_ast(py_ast)
-    if value is None:
-        return "Invalid Expression: Not a qubit literal or Tensor"
-    tensor = qwerty_ast.NodeBox.new_qubit_tensor(py_ast.body[0].value.value)
-    return tensor.get_tensor()
 
-def unpack_ast(ast):
+    if not isinstance(py_ast, (ast.Interactive, ast.Module)):
+        return "Invalid AST"
+    
+    if (len(py_ast.body) == 1):
+        value = py_ast.body[0]
+    else:
+        return "Invalid body; length of body is not 1"
+    
+    if not isinstance(value, ast.Expr):
+        return "Invalid syntax; expected an expression"
+    
     try:
-        return ast.body[0].value.value
-    except AttributeError:
-        return None
+        return convert_expr(value.value)
+    
+    except Exception as e:
+        raise QwertySyntaxError("Just in case some random things gets thru") from e    
+#   1) check if the instance is a binOp, if so, check its left or right values and then perform the binOp. only sum is implemented for now
+#   2) check if the instance is a constant and return it as a tensor, return value in a tensor
+
+    
+def convert_expr(node):
+    if isinstance(node, ast.Constant):
+        val = node.value
+        if isinstance(val, str) and all(num in "01" for num in val):
+            tensor = qwerty_ast.NodeBox.new_qubit_tensor(node.value)         
+            return tensor
+        else:
+            raise QwertySyntaxError(f"Invalid character/value: {val}, e.g. not 1s and 0s")
+        
+    elif isinstance(node, ast.BinOp):
+        left = convert_expr(node.left)
+        right = convert_expr(node.right)
+        if isinstance(node.op, ast.Mult):
+            return concat_tensors(left, right)
+        else:
+            raise NotImplementedError("Operation is not supported, only addition for now")
+        
+    else:
+        raise QwertySyntaxError(f"Unsupported AST node: {type(node).__name__}")
+        
+    # elif isinstance(node, ast.Name):
+    #     symbol_table[node.id] = qwerty_ast.NodeBox.resolve_name(node.id)
+    #     return qwerty_ast.NodeBox.resolve_name(node.id)
+    # else:
+    #     raise NotImplementedError("Variable not supported")
+
+def concat_tensors(left, right):
+    left_val = left.get_tensor()
+    right_val = right.get_tensor()
+    combined = str(left_val) + str(right_val)
+    return qwerty_ast.NodeBox.new_qubit_tensor(combined)
```

Comment:
I don't think we should do this stuff with strings here because Qwerty allows taking the tensor product of things that are not strings, like bits, functions, and bases.

When we encounter `e1 * e2`, we need to parse `e1` into a Qwerty AST node, then parse `e2` into a Qwerty AST node, then finally construct a `Tensor` node that contains both of those two nodes. It looks like `new_tensor()` does the last part well. That way, your code is consistent no matter what kind of thing is being tensored together.

Reply:
I'm going to write some kind of "AST canonicalization" in Dinesh's branch that flattens nested `Tensor` nodes. That canonicalization would guarantee that `'0' * '1' * '0'` is always represented by a `Tensor` node with 3 children like this:
```
           Tensor
         /   |   \
        /    |    \
ZeroQubit OneQubit ZeroQubit
```
 instead of a Tensor node with a Tensor child, like this:
```
         Tensor
         /    \
        /      \
     Tensor    ZeroQubit 
     /   \
    /     \
ZeroQubit OneQubit
```

But your code can produce the latter form, no problem, and if I port that canonicalization code over here when I'm done, then that canonicalization step can run before evaluation and convert the second form into the first.